types/tokenizr/index.d.ts

```diff
@@ -0,0 +1,268 @@
+
+    /**
+     * Get/set state
+     */
+    state(): string;
+    /**
+     * Replaces the current state
+     */
+    state(state: string): this;
+
+    /**
+     * Set a tag
+     */
+    tag(tag: string): this;
+
+    /**
+     * Check whether tag is set
+     */
+    tagged(tag: string): boolean;
+
+    /**
+     * Determine and return next token
+     */
+    token(): Token | null;
+
+    /**
+     * Determine and return all tokens
+     */
+    tokens(): Token[];
+
+    /**
+     * Unset a tag
+     */
+    untag(tag: string): this;
+
+    static readonly ParsingError: typeof ParsingError;
+    static readonly ActionContext: typeof ActionContext;
+    static readonly Token: typeof Token;
+}
+
+type Action = (
+    this: ActionContext,
+    ctx: ActionContext,
+    found: RegExpExecArray,
+    rule: {
+        state: string;
+        pattern: RegExp;
+        action: RuleAction;
+        name: string;
+    }
+) => void;
+
+type RuleAction = (
+    this: ActionContext,
+    ctx: ActionContext,
+    found: RegExpExecArray
+) => void;
+
+declare class ActionContext {
+    constructor(e: any);
+
+    /**
+     * Accept current matching as a new token
+     */
+    accept(type: string, value?: any): this;
+
+    /**
+     * Store and retrieve user data attached to context
+     */
+    data(key: string, value?: any): any;
+
+    /**
+     * Mark current matching to be ignored
+     */
+    ignore(): this;
+
+    /**
+     * Retrieve information of current matching
+     */
+    info(): { line: number; column: number; pos: number; len: number };
+
+    /**
+     * Pop state
+     */
+    pop(): string;
+
+    /**
+     * Push state
+     */
+    push(state: string): this;
+
+    /**
+     * Rark current matching to be rejected
+     */
+    reject(): this;
+
+    /**
+     * Mark current matching to be repeated from scratch
+     */
+    repeat(): this;
+
+    /**
+     * Get/set state
+     */
+    state(): string;
+    /**
+     * Replaces the current state
+     */
+    state(state: string): this;
+
+    /**
+     * Immediately stop tokenization
+     */
+    stop(): this;
+
+    /**
+     * Set a tag
+     */
+    tag(tag: string): this;
+
+    /**
+     * Check whether tag is set
+     */
+    tagged(tag: string): boolean;
+
+    /**
+     * Unset a tag
+     */
+    untag(tag: string): this;
+}
+
+declare class ParsingError extends Error {
+    constructor(
+        message: string,
+        pos: number,
+        line: number,
+        column: number,
+        input: string
+    );
+
+    /**
+     * Render a useful string representation
+     */
+    toString(): string;
+}
+
+declare class Token {
+    constructor(
+        type: string,
+        value: any,
+        text: string,
+        pos?: number,
+        line?: number,
+        column?: number
+    );
+
+    isA(type: string, value?: any): boolean;
+
+    /**
+     * Render a useful string representation
+     */
+    toString(): string;
+}
+
+export = Tokenizr;
```

Comment:
this looks good, but i noticed the original library [exposes some of those classes as properties of Tokenizr](https://github.com/rse/tokenizr/blob/master/src/tokenizr.js#L708). how are you exporting those properties?

Reply:
They are declared on the Tokenizr class, lines 139 - 141.

Reply:
ok, i see them now. thanks!